Fix #15064 - Export SQL option hidden on sql query export results

saurass opened the pull request:
# Description 
Fixes #15064 - Export SQL option hidden on sql query export results.

* Shows export as SQL option for queries that do not involve use of table.
* In case of such queries the raw query (entered by user) will be exported as *.sql file.

# Approach
* To recognize this kind of query (which do not use any DB or table) we need a flag to be passed
In `libraries/classes/Display/Results.php` added `free_query` parameter.
* This adds `free_query` parameter to the export link.
* Now this parameter is passed in the check in `libraries/classes/Plugins/Export/ExportSql.php` this makes the SQL option visible for export for these type of queries.
* Now further we need to pass this parameter to `export.php`
* So made changes to `templates/display/export/hidden_inputs.twig` and added a hidden field as `free_query`
* At the moment none of the export plugins are designed to dump raw query (as entered by user), so made changes to that.
* made a function `exportRawQuery()` in `libraries/classes/Plugins/Export/ExportSql.php`
* Now for this part function to be executed, we define `$whatStrucOrData` to `free_query` in `export.php`
* This way we dump the raw queries in sql dump file.

# Result
* Video illustrating the progress https://drive.google.com/open?id=15zS1U24ids9TFOvCe9Shi6p7DIn25p2d

# Further works
* Now I have implemented the `exportRawQuery()` only in `ExportSql.php` plugin, we need to implement this function for all other export plugins.

# Contribution Checklist
- [x] Make sure you have read our [CONTRIBUTING.md](https://github.com/phpmyadmin/phpmyadmin/blob/master/CONTRIBUTING.md) document.
- [x] Make sure you are making a pull request against the correct branch. For example, for bug fixes in a released version use the corresponding QA branch and for new features use the _master_ branch. If you have a doubt, you can ask as a comment in the bug report or on the mailing list.
- [x] Every commit has proper `Signed-off-by` line as described in our [DCO](https://github.com/phpmyadmin/phpmyadmin/blob/master/DCO). This ensures that the work you're submitting is your own creation.
- [x] Every commit has a descriptive commit message.
- [x] Every commit is needed on its own, if you have just minor fixes to previous commits, you can squash them.
- [x] Any new functionality is covered by tests.

saurass commented:
@williamdes @ibennetch please review and suggest changes : )

saurass commented:
@williamdes @mauriciofauth 

I found that the SQL option isn't coming because it is not initialized from `libraries/classes/Plugins/Export/ExportSql.php` on line 72 to 79

## Reason
* When we are calling to export any table, we are sending `single_table` as parameter (boolean).
* In case we export a table we are setting it to true.
* But in case of join queries or independent union queries (as in the issue mentioned by William) we do not send single_table and it is set to false and thus SQL option is not initialized.

## Possible Solutions
* Solution could be that when we enter a query that doesn't involve any table, we can pass a new param like `free_query` and add this to enable the SQL option and the user can further download the SQL file for it.

## Work So far
I have made following progress till now-
* Initializing a `free_query` parameter in case queries do not require any table or database.
* Use this to show the SQL option while exporting.
* Again this is used to export the raw query
* Fixed any problem with mime_type, (already working)
* When exporting such queries, the default file name will be like `export.extension`

Video illustrating progress https://drive.google.com/open?id=1oQeBUCpXLGWxIbZI_zP2BH9G-W3sM7RG

Please lemme know if I am going in the right direction.

Thank You : )

saurass commented:
Now I am working on removing DB and Table comments in the file that are exported. Will update soon : )

williamdes commented:
@sudo-bot Deploy pr

saurass committed:
Fix 15064 - instead of using free_query parameter, now updating export_type to raw -- more modular approach

Signed-off-by: Saurabh Srivastava <saurabhsrivastava312@gmail.com>

saurass commented:
I have just made few more changes @williamdes . Please review this and try to deploy.
Hope this goes well : )

williamdes commented:
@sudo-bot Deploy pr

@saurass you have Deploy access, so you can ask the bot to do so :)

saurass commented:
Okay that sounds cool : )

saurass commented:
@sudo-bot Deploy pr

saurass commented:
@williamdes , The bot seems to be unhappy, any idea what's going wrong?

williamdes commented:
@saurass I am fixing the infrastructure, I have memory issues with `composer install`
I will re trigger a build when fixed ;)

saurass commented:
updates...

## Approach
#### After clicking GO to execute query
* `import.php` is called which contains `Export` link
![tempsnip](https://user-images.githubusercontent.com/31488093/55878045-3eab4580-5bb9-11e9-80ea-4e14980d750a.png)
* This link must contain a parameter to identify that the query doesn't require use of table or database. This parameter will be passed to `tbl_export.php`
* So in `libraries/classes/Display/Results.php` initialize new parameter `raw_query` in case no table is selected. (see line 4964)
* This will give us a parameter `raw_query` when `tbl_export.php` is called (Export is clicked)

#### After export in called (when opening `tbl_export.php`)
* We use the above initialized parameter `raw_query` to show the SQL option
* So in `libraries/classes/ExportSql.php` we use the parameter `raw_query` to unhide the SQL option. (See line 81)
* In `tbl_export.php` we initialize `$export_type` and set it to "raw" if `raw_query` parameter is passed. (see line 138)
* Now we are able to see SQL option as shown below -
![Capture](https://user-images.githubusercontent.com/31488093/55878528-415a6a80-5bba-11e9-9aa2-ab285d191808.PNG)

#### After we click GO to export in SQL format. (`export.php` will be called)
* We need to export the raw query as the user inputs.
* So, we set `$err_url` as `server_sql.php` in case any error occurs, redirect to this page. (see line 289 `export.php`)
* In `export.php` we call `Export::exportRaw()` in case the `$export_type == 'raw'` (see line 477)
* Now we make static function `exportRaw()` in `libraries/classes/Export.php`. (see line 796)
* This function further calls `exportRawQuery()` function of export plugin (here `ExportSql` plugin) [This function has to be implemented in all types of export plugins]
* Now in `libraries/classes/Plugins/Export/ExportSql.php` we define function `exportRawQuery()` (see line 1969)


This downloads a file names `localhost.sql`

saurass commented:
@williamdes I think this PR will work well while exporting SQL queries.

saurass commented:
This should also work well with JSON export now.

saurass commented:
@sudo-bot Deploy pr

saurass commented:
Export CSV now working. : )

williamdes commented on libraries/classes/Export.php:
```diff
@@ -858,6 +866,33 @@ public function exportDatabase(
+    /**
+     * Export raw query
+     *
+     * @param $whatStrucOrData
```
please add parameter types :)
eg: `@param string $parameter`

saurass replied:
okay

williamdes commented on libraries/classes/Plugins/Export/ExportJson.php:
```diff
@@ -287,4 +287,75 @@ public function exportData(
+        $GLOBALS['dbi']->freeResult($result);
+
+        return true;
+        return $this->exportData('','','', $err_url, $sql_query);
```
this line will never be reached

saurass replied:
I forgot to remove the last line. It is not required. Removing : )

williamdes commented on libraries/classes/Plugins/Export/ExportSql.php:
```diff
@@ -1997,6 +1997,21 @@ private function _getTableComments(
+    {
+        $dump = '';
+        $dump .= $sql_query;
+        return $this->export->outputHandler($dump);
```
You can use `$sql_query` instead of `$dump`

saurass replied:
Yeah cool.

williamdes commented on libraries/classes/Plugins/Export/ExportSql.php:
```diff
@@ -1997,6 +1997,21 @@ private function _getTableComments(
+     *
+     * @param string $err_url   the url to go back in case of error
+     * @param string $sql_query the rawquery to output
+     *
```
missing parameter `$crlf`

saurass replied:
okay

williamdes commented on tbl_export.php:
```diff
@@ -145,15 +145,15 @@
 if (! isset($multi_values)) {
     $multi_values = '';
 }
+if (isset($_REQUEST['raw_query'])) {
```
if possible use `_GET` or `_POST`

saurass replied:
$_POST can be used for sure, but not sure if it is used anywhere else in the code base.

saurass replied:
Okay, I will get this to POST

williamdes reviewed (changes requested):
The code looks okay, and works fine on deployed instance

saurass commented:
@sudo-bot Deploy pr

williamdes commented on export.php:
```diff
@@ -374,6 +376,12 @@
 
+// For raw query export, filename will be export.extension
+// TODO: It can be make dynamic, user may enter name of file.
+if($export_type == 'raw') {
```
please add a space before `(`

saurass replied:
Yeah sure

saurass commented:
I have updated few more plugins, they must work well.

I am working on that for `Codegen` `LaTeX` and `XML` . I may require help in these

saurass commented:
@sudo-bot Deploy pr

mauriciofauth commented on libraries/classes/Controllers/Server/SqlController.php:
```diff
@@ -19,16 +19,18 @@
      * @return string HTML
      */
-    public function index(): string
+    public function index($query_to_show): string
```
We are using an array of request parameters in the actions of the controllers. Example:
https://github.com/phpmyadmin/phpmyadmin/blob/13ef404d1e0fddc4c8595992b0385d2cb830a9a3/libraries/classes/Controllers/BrowseForeignersController.php#L47-L51

saurass replied:
Okay, I will pass it as an array. Thanks

MauricioFauth reviewed (changes requested):
Hi, @saurass. Thanks for your contribution.

You need to fix the phpstan error [1] and the phpcs errors [2].

- [1] https://travis-ci.org/phpmyadmin/phpmyadmin/jobs/519224641#L631
- [2] https://travis-ci.org/phpmyadmin/phpmyadmin/jobs/519224640#L641

saurass commented:
Travis-ci builds are failing any ideas why?

cc @williamdes @mauriciofauth

saurass commented:
@williamdes @mauriciofauth 
I have made changes to all the ExportPlugins except "CodeGen" of which I have no idea.
So, should we work with the codegen part , if yes please help me get that done.

Otherwise I think this PR should work well (is ready I guess). Please test this and lemme know any thing that bothers you or is not working well.

Thank You : )

saurass commented:
@sudo-bot /deploy

saurass commented:
@sudo-bot Deploy pr

mauriciofauth commented:
> Travis-ci isn't working any ideas why?

@saurass Can you rebase against `master` and squash your commits?

saurass commented:
> > Travis-ci isn't working any ideas why?
> 
> @saurass Can you rebase against `master` and squash your commits?

Thank you @mauriciofauth

saurass commented:
@sudo-bot Deploy pr

mauriciofauth commented on export.php:
```diff
@@ -374,6 +376,12 @@
 }
 
+// For raw query export, filename will be export.extension
+// TODO: It can be make dynamic, user may enter name of file.
```
Could you remove this TODO?

saurass replied:
Ya sure, I will remove this.

saurass commented:
@mauriciofauth @williamdes please suggest further changes.

williamdes commented:
@saurass sorry for not reviewing, can you please fix the conflicts ?

saurass commented:
> @saurass sorry for not reviewing, can you please fix the conflicts ?

You mean the merge conflicts. Right?

williamdes commented:
> > @saurass sorry for not reviewing, can you please fix the conflicts ?
> 
> You mean the merge conflicts. Right?

Yes

williamdes commented:
@saurass I rebased and add types to your great work, thank your for your contribution !

williamdes commented:
@sudo-bot Deploy PR

williamdes commented:
@ibennetch and @mauriciofauth could I have your review on this old PR that should be merged a long time ago ?

williamdes committed:
Some code fixes

Signed-off-by: William Desportes <williamdes@wdes.fr>

williamdes commented:
CI should pass, testing in progress
Maybe I will add some test coverage before merging

williamdes commented:
Issues found:
- [ ] - Export as visible text, go back: could be better

williamdes committed:
Add a message when the plugin does not support the export method

Example:
- no db, table selected.
- Type a raw query, export using codegen.
- See error message
CodeGen can not export if it does not have db and table params

Signed-off-by: William Desportes <williamdes@wdes.fr>

williamdes committed:
Support return back from raw export

Signed-off-by: William Desportes <williamdes@wdes.fr>

williamdes committed:
Move some code to getFinalFilenameAndMimetypeForFilename and add tests

Signed-off-by: William Desportes <williamdes@wdes.fr>

williamdes commented:
@victorphoenix3 Could you help me review this PR please ?
cc @kartik1000

williamdes commented:
cc @boherm

kartik1000 commented:
Hey @williamdes, Is the code change made in  templates/display/export/option_header.twig done to inform the user that we are exporting a raw query. I didn't understand its use much. Can you explain?

williamdes commented:
> Hey @williamdes, Is the code change made in templates/display/export/option_header.twig done to inform the user that we are exporting a raw query. I didn't understand its use much. Can you explain?

Hey @kartik1000 
You are right, it is only to inform the user that we are exporting a raw query and not a db or a table.

williamdes commented:
Thank you for your hard work and patience !
Merged :tada:

saurass commented:
> Thank you for your hard work and patience !
> Merged

Thanks a lot : )

williamdes commented on libraries/classes/Plugins/Export/ExportSql.php:
```diff
@@ -2006,6 +2013,20 @@ private function _getTableComments(
+     */
+    public function exportRawQuery(string $err_url, string $sql_query, string $crlf): bool
+    {
+        return $this->export->outputHandler($sql_query);
```
This is a mistake, fixed by d258e4ed142ca69efd0f68e9018529cb11f9d2f9
